django/db/models/sql/compiler.py

```diff
@@ -903,17 +905,102 @@ def __init__(self, *args, **kwargs):
         self.return_id = False
         super(SQLInsertCompiler, self).__init__(*args, **kwargs)
 
-    def placeholder(self, field, val):
+    def field_as_sql(self, field, val):
+        """
+        Take a field, and a value intended to be saved on that field, and
+        return placeholder SQL and accompanying params. Checks for raw values,
+        expressions and fields with get_placeholder() defined in that order.
+        """
+
         if field is None:
             # A field value of None means the value is raw.
-            return val
+            sql, params = val, []
+
+        elif hasattr(val, 'as_sql'):
+            # This is an expression, let's compile it.
+            sql, params = self.compile(val)
+
         elif hasattr(field, 'get_placeholder'):
             # Some fields (e.g. geo fields) need special munging before
             # they can be inserted.
-            return field.get_placeholder(val, self, self.connection)
+            sql, params = field.get_placeholder(val, self, self.connection), [val]
```

Comment:
What happens in the case where field defines get_placeholder() and the val is an expression? Do we want to support this (as IIRC this is Oracle GIS specific case)?